### PR TITLE
prometheus and alertmanager: added imagePullSecrets support to statef…

### LINKED ISSUE #315: Add ImagePullSecrets to support private registries
When specifying a baseImage in a [Prometheus TPR](https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#prometheusspec) and in an [Alertmanager TPR](https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#alertmanagerspec) there is no way to provide  [ImagePullSecrets](https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod) as in [kube pod spec,](https://kubernetes.io/docs/api-reference/v1.6/#podspec-v1-core) to authenticate with private secure registries.

As a consequence, Operator cannot manage Prometheus and Alertmanagers if their images are not maintained in public registries. This is a limitation, if your environment is not open to the public internet, or if you simply want to maintain your own images in a private registry.

 I am suggesting to extend the TPR spec to support optional ImagePullSecrets and if specified, include them in Prometheus and Alertmanager StatefulSets on creation and update.

PR is coming

### PR DESCRIPTION
**Pull request by nirdothan**
…ulsets

This enhancement enables users to maintain Prometheus and Alertmanager
images in private/authenticated image registries

Submitted on behalf of Amdocs Development Limited

Fixes #315

### CODE REVIEW

**Comment by brancz** [comment]
Test this please.

File: pkg/prometheus/statefulset.go [hunk 1]
```diff
@@ -86,6 +86,12 @@ func makeStatefulSet(p v1alpha1.Prometheus, old *v1beta1.StatefulSet, config *Co
 		},
 		Spec: *spec,
 	}
+
+	if p.Spec.ImagePullSecrets != nil && len(p.Spec.ImagePullSecrets) > 0 {
+		statefulset.Spec.Template.Spec.ImagePullSecrets = make([]v1.LocalObjectReference, len(p.Spec.ImagePullSecrets))
```
Comment (brancz) [inline pkg/prometheus/statefulset.go]: Is there a specific reason for the copying here that I'm missing? If not just do 

```
statefulset.Spec.Template.Spec.ImagePullSecrets = p.Spec.ImagePullSecrets
```

Comment (nirdothan) [inline pkg/prometheus/statefulset.go]: I figured it would be safer not to keep a reference the the cache memory, but hey, you're the boss ;)

Comment (brancz) [inline pkg/prometheus/statefulset.go]: Don't think this actually makes a difference here.

**Comment by nirdothan** [comment]
replaced the copy with direct assignments

**Comment by brancz** [comment]
lgtm. Thanks a lot for reporting and contributing!! :tada:

**Comment by nirdothan** [comment]
My pleasure  🎉

**Comment by bamb00** [comment]
Hi,
Is there documentation showing how to manage the alertmanager statefulset image in a private docker registry? I'm looking for where I can add the "imagePullSecret:" field to the template 
Thanks.

**Comment by nirdothan** [comment]
Hi, see https://github.com/coreos/prometheus-operator/blob/
master/Documentation/api.md#AlertmanagerSpec

It's used the same way you use it in pod template. i.e.
  imagePullSecrets:
  - name: <secret name>

See
https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/

**Comment by bamb00** [comment]
Hi,

I use the alertmanager image from a private registry according to alertmanager spec and is not working. 

There's also a second image config-reloader images use in the alertmanager-main.  How do I use it?
**--config-reloader-image=quay.io/coreos/configmap-reload:v0.0.1**

I use the AlertManager Spec, https://github.com/coreos/prometheus-operator/blob/5da727d9f98e9644d01a17b3c08eb8a68f793533/Documentation/api.md#alertmanagerspec

What am I missing? This seems to work for prometheus.

**Private Registry version:**
apiVersion: monitoring.coreos.com/v1
kind: Alertmanager
metadata:
  name: main
  labels:
    alertmanager: main
  namespace: monitoring
spec:
  replicas: 3
  version: 0.9.1.0.6412fc7f
  baseImage: myprivate_registry_hub/docker-dev/alertmanager
  imagePullSecrets:
  - name: tricregkey

**Original from Manifests**: 
apiVersion: monitoring.coreos.com/v1
kind: Alertmanager
metadata:
  name: main
  labels:
    alertmanager: main
  namespace: monitoring
spec:
  replicas: 3
  version: v0.9.1

Thanks in Advance.

**Comment by brancz** [comment]
@bamb00 in your case you want to set the default base image and config reloader images when you start your prometheus operator. The config reloader is an implementation detail, which is why we don't have it as a field of the Alertmanager or Prometheus resource.

**Comment by brancz** [comment]
yes, just point those base image flags to your private registry

**Comment by bamb00** [comment]
Thanks.

**Comment by bamb00** [comment]
@brancz, Still not able to get alertmanager to deploy to k8s using private registry.  But the prometheus-k8s deployed using private registry.  The prometheus-operator

This is what I have in alertmanager.  What am I missing for alertmanager?

        apiVersion: monitoring.coreos.com/v1
        kind: Alertmanager
        metadata:
          name: main
          labels:
            alertmanager: main
          namespace: monitoring
        spec:
          replicas: 3
          version: 0.9.1.0.6412fc7f            #with and without the double-quotes
          imagePullSecrets:
          - name: myregkey

Logs:
       ts=2018-01-30T21:30:51Z caller=operator.go:321 component=alertmanageroperator msg="Alertmanager updated" key=monitoring/main
ts=2018-01-30T21:30:51Z caller=operator.go:377 component=alertmanageroperator msg="sync alertmanager" key=monitoring/main
E0130 21:30:51.841015       1 operator.go:253] Sync "monitoring/main" failed: making the statefulset, to create, failed: parse version: Invalid character(s) found in patch number "1.0.6412fc7f"

**Comment by brancz** [comment]
It's not a bug. Versions/tags are expected to follow semantic versioning. I highly recommend to just push the upstream images to your own registry. If this is not an option, then make sure to tag it following semantic versioning. I believe in your case what would make sense then, is `0.9.1-0.6412fc7f`.

**Pull request merged**